File: feature/locations/src/main/java/ru/example/gnt/locations/presentation/details/recyclerview/CharacterViewHolder.java
package ru.example.gnt.locations.presentation.details.recyclerview;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.RequestManager;

import ru.example.gnt.common.model.characters.CharacterListItem;
import ru.example.gnt.ui.databinding.CharacterItemBinding;

public class CharacterViewHolder extends RecyclerView.ViewHolder {

    protected CharacterItemBinding binding;
    private final CharacterViewHolderEventListener listener;
    private final RequestManager glide;

    public CharacterViewHolder(@NonNull CharacterItemBinding binding, CharacterViewHolderEventListener listener, RequestManager glide) {
        super(binding.getRoot());
        this.listener = listener;
        this.binding = binding;
        this.glide = glide;
    }

    public void onBind(CharacterListItem item) {
        try {
            binding.tvGender.setText(item.getGender().getValue());
            binding.tvStatus.setText(item.getStatus().getValue());
            binding.tvStatus.setBackgroundColor(item.getStatus().getColor().getValue(binding.getRoot().getContext()).getDefaultColor());
        } catch (Exception ignored) {}
        binding.tvName.setText(item.getName());
        binding.tvSpecies.setText(item.getSpecies());
        binding.tvStatus.setWidth(binding.ivCharacter.getMaxWidth());
        binding.getRoot().setOnClickListener(v -> {
            listener.onItemClicked(item.getId());
        });
        glide.load(item.getImage()).into(binding.ivCharacter);

    }

}

File: feature/episodes/src/main/java/ru/example/gnt/episodes/presentation/episode_details/recyclerview/CharacterViewHolder.kt
package ru.example.gnt.episodes.presentation.episode_details.recyclerview

import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.bumptech.glide.Glide
import com.bumptech.glide.RequestManager
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.target.Target
import ru.example.gnt.common.model.characters.CharacterListItem
import ru.example.gnt.episodes.databinding.CharacterItemBinding

class CharacterViewHolder(
    private val binding: CharacterItemBinding,
    private val onItemClicked: (id: Int) -> Unit,
    private val glide: RequestManager
) : ViewHolder(binding.root) {

    fun onBind(item: CharacterListItem?) {
        if (item != null) {
            with(binding) {
                setImage(item)
                tvStatus.width = ivCharacter.maxWidth
                tvName.text = item.name
                tvGender.text = item.gender?.value
                tvSpecies.text = item.species
                tvStatus.text = item.status?.value
                item.status?.color?.getValue(binding.root.context)?.defaultColor?.let {
                    tvStatus.setBackgroundColor(
                        it
                    )
                }
                root.setOnClickListener {
                    onItemClicked.invoke(item.id)
                }
            }
        }
    }

    private fun setImage(item: CharacterListItem) {
        glide
            .load(item.image)
            .addListener(object : RequestListener<Drawable> {
                override fun onLoadFailed(
                    e: GlideException?,
                    model: Any?,
                    target: Target<Drawable>?,
                    isFirstResource: Boolean
                ): Boolean {

                    return false
                }

                override fun onResourceReady(
                    resource: Drawable?,
                    model: Any?,
                    target: Target<Drawable>?,
                    dataSource: DataSource?,
                    isFirstResource: Boolean
                ): Boolean {
                    return false
                }
            })
            .into(binding.ivCharacter)
    }

    companion object {
        operator fun invoke(
            parent: ViewGroup,
            onItemClicked: (id: Int) -> Unit,
            glide: RequestManager
        ) : CharacterViewHolder {
            val holder =   CharacterViewHolder(
                binding = CharacterItemBinding.inflate(
                    LayoutInflater.from(parent.context),
                    parent,
                    false
                ),
                onItemClicked = onItemClicked,
                glide = glide
            )
            (holder.binding.root.layoutParams as StaggeredGridLayoutManager.LayoutParams).apply {
                width = parent.context.resources.displayMetrics.widthPixels / 2
            }.also {
                holder.binding.root.layoutParams = it
            }
            return holder
        }
    }
}
